src/NServiceBus.Gateway.RavenDB.AcceptanceTests/GatewayTestSuiteConstraints.cs

```diff
@@ -8,55 +8,64 @@
 
 namespace NServiceBus.Gateway.AcceptanceTests
 {
+    using System.Collections.Concurrent;
+
     public partial class GatewayTestSuiteConstraints
     {
-        public Task ConfigureDeduplicationStorage(string endpointName, EndpointConfiguration configuration, RunSettings settings)
+        public Task<GatewayDeduplicationConfiguration> ConfigureDeduplicationStorage(string endpointName, EndpointConfiguration configuration, RunSettings settings)
         {
-            var ravenGatewayDeduplicationConfiguration = new RavenGatewayDeduplicationConfiguration((builder, _)=> 
+            var ravenGatewayDeduplicationConfiguration = new RavenGatewayDeduplicationConfiguration((builder, _) =>
             {
-                databaseName = Guid.NewGuid().ToString();
+                var databaseName = Guid.NewGuid().ToString();
                 var documentStore = GetInitializedDocumentStore(databaseName);
 
                 documentStore.Maintenance.Server.Send(new CreateDatabaseOperation(new DatabaseRecord(databaseName)));
 
+                databases.Add(databaseName);
+
                 return documentStore;
             });
 
             var gatewaySettings = configuration.Gateway(ravenGatewayDeduplicationConfiguration);
             configuration.GetSettings().Set(gatewaySettings);
 
-            return Task.CompletedTask;
+            return Task.FromResult<GatewayDeduplicationConfiguration>(ravenGatewayDeduplicationConfiguration);
         }
 
         public async Task Cleanup()
         {
-            // Periodically the delete will throw an exception because Raven has the database locked
-            // To solve this we have a retry loop with a delay
-            var triesLeft = 3;
-
-            while (triesLeft-- > 0)
+            while (!databases.IsEmpty)
             {
-                try
+                // Periodically the delete will throw an exception because Raven has the database locked
+                // To solve this we have a retry loop with a delay
+                var triesLeft = 3;
+
+                while (triesLeft-- > 0)
                 {
-                    // We are using a new store because the global one is disposed of before cleanup
-                    using (var storeForDeletion = GetInitializedDocumentStore(databaseName))
+                    try
                     {
-                        storeForDeletion.Maintenance.Server.Send(new DeleteDatabasesOperation(storeForDeletion.Database, hardDelete: true));
-                        break;
+                        if (databases.TryTake(out var databaseName))
```

Comment:
if the deletion fails, the value would still be removed from the bag? In this case, we wouldn't attempt to delete this again?

Reply:
yeah, right. Using the semaphore slim was simpler 😉

Reply:
as said, we can keep the semaphore approach if it's simpler, I have no objections.